test/e2e/upgrades/serviceaccount_admission_controller_migration.go

```diff
@@ -0,0 +1,135 @@
+	podAfterMigrationName  = "pod-after-migration"
+)
+
+// ServiceAccountAdmissionControllerMigrationTest test that a pod is functioning before and after
+// a cluster upgrade.
+type ServiceAccountAdmissionControllerMigrationTest struct {
+	pod *v1.Pod
+}
+
+// Name returns the tracking name of the test.
+func (ServiceAccountAdmissionControllerMigrationTest) Name() string {
+	return "[sig-auth] serviceaccount-admission-controller-migration"
+}
+
+// Setup creates pod-before-migration which has legacy service account token.
+func (t *ServiceAccountAdmissionControllerMigrationTest) Setup(f *framework.Framework) {
+	t.pod = createPod(f, podBeforeMigrationName)
+	inClusterClientMustWork(f, t.pod)
+}
+
+// Test waits for the upgrade to complete, and then verifies pod-before-migration
+// and pod-after-migration are able to make requests using in cluster config.
+func (t *ServiceAccountAdmissionControllerMigrationTest) Test(f *framework.Framework, done <-chan struct{}, upgrade UpgradeType) {
+	ginkgo.By("Waiting for upgrade to finish")
+	<-done
+
+	ginkgo.By("Starting post-upgrade check")
+	podBeforeMigration, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(context.TODO(), podBeforeMigrationName, metav1.GetOptions{})
+	framework.ExpectNoError(err)
+	if podBeforeMigration.GetUID() != t.pod.GetUID() {
+		framework.Failf("Pod %q GetUID() = %q, want %q.", podBeforeMigration.Name, podBeforeMigration.GetUID(), t.pod.GetUID())
+	}
+	if podBeforeMigration.Status.ContainerStatuses[0].RestartCount != 0 {
+		framework.Failf("Pod %q RestartCount = %d, want 0.", podBeforeMigration.Name, podBeforeMigration.Status.ContainerStatuses[0].RestartCount)
+	}
+	if len(podBeforeMigration.Spec.Volumes) != 1 || podBeforeMigration.Spec.Volumes[0].Secret == nil {
+		framework.Failf("Pod %q Volumes[0].Secret = nil, want non-nil.", podBeforeMigration.Name)
+	}
+
+	ginkgo.By("Checking pod-after-migration makes successful requests using in cluster config")
+	podAfterMigration := createPod(f, podAfterMigrationName)
+	if len(podAfterMigration.Spec.Volumes) != 1 || podAfterMigration.Spec.Volumes[0].Projected == nil {
+		framework.Failf("Pod %q Volumes[0].Projected.Sources = nil, want non-nil.", podAfterMigration.Name)
+	}
+	inClusterClientMustWork(f, podAfterMigration)
+
+	ginkgo.By("Finishing post-upgrade check")
+}
+
+// Teardown cleans up any remaining resources.
+func (t *ServiceAccountAdmissionControllerMigrationTest) Teardown(f *framework.Framework) {
+	// rely on the namespace deletion to clean up everything
+}
+
+func inClusterClientMustWork(f *framework.Framework, pod *v1.Pod) {
+	var logs string
+	if err := wait.Poll(1*time.Minute, 20*time.Minute, func() (done bool, err error) {
+		framework.Logf("Polling logs")
+		logs, err = e2epod.GetPodLogs(f.ClientSet, pod.Namespace, pod.Name, "inclusterclient")
+		if err != nil {
+			framework.Logf("Error pulling logs: %v", err)
+			return false, nil
+		}
+		_, err = e2eauth.ParseInClusterClientLogs(logs)
```

Comment:
if we have a single-server cluster, network request errors during the upgrade are expected. we specifically want to fail if this encounters an unauthorized error, right?

Reply:
I'm not sure the existing ParseInClusterClientLogs method is what we want here... we want to verify:
1. no unauthorized (401) errors are seen (ParseInClusterClientLogs fails on any error)
2. at least one successful authenticated request was made after a specific time, possibly using the SinceTime pod log option (ParseInClusterClientLogs scans all logs, which means we can't distinguish a successful request before and after the upgrade):
  * in setup, after the pod creation
  * after upgrade, after the API server restarts

Reply:
yes, there are bunch of network errors when testing `pod-before-migration` after upgrade as expected. to only test on 401/403 we probably need to change the inclusterclient to report the HTTP code of err in the log and rebuild the image. the stronger assertion that there is no failure since upgrade is sufficient right? the only issue is that it might fail for other reasons as well. WDYT?